server/app/services/applicant/ApplicantService.java

```diff
@@ -1045,4 +1050,118 @@ abstract static class Builder {
+            });
+  }
+
+  /** Maps address suggestion and corrected state into a form data compatible map */
+  private ImmutableMap<String, String> buildCorrectedAddressAsFormData(
+      AddressQuestion addressQuestion,
+      Optional<AddressSuggestion> suggestionMaybe,
+      String selectedAddress) {
+
+    Map<String, String> questionPathToValueMap = new HashMap<>(Map.of());
+
+    if (suggestionMaybe.isPresent()) {
+      AddressSuggestion suggestion = suggestionMaybe.get();
+      Address address = suggestion.getAddress();
+      AddressLocation location = suggestion.getLocation();
+
+      questionPathToValueMap.put(addressQuestion.getStreetPath().toString(), address.getStreet());
+      questionPathToValueMap.put(addressQuestion.getLine2Path().toString(), address.getLine2());
+      questionPathToValueMap.put(addressQuestion.getCityPath().toString(), address.getCity());
+      questionPathToValueMap.put(addressQuestion.getStatePath().toString(), address.getState());
+      questionPathToValueMap.put(addressQuestion.getZipPath().toString(), address.getZip());
+      questionPathToValueMap.put(
+          addressQuestion.getLatitudePath().toString(), location.getLatitude().toString());
+      questionPathToValueMap.put(
+          addressQuestion.getLongitudePath().toString(), location.getLongitude().toString());
+      questionPathToValueMap.put(
+          addressQuestion.getWellKnownIdPath().toString(), location.getWellKnownId().toString());
+      questionPathToValueMap.put(
+          addressQuestion.getCorrectedPath().toString(),
+          CorrectedAddressState.CORRECTED.toString());
+    } else if (selectedAddress.equals(AddressCorrectionBlockView.USER_KEEPING_ADDRESS_VALUE)) {
+      questionPathToValueMap.put(
+          addressQuestion.getCorrectedPath().toString(),
+          CorrectedAddressState.AS_ENTERED_BY_USER.toString());
+    } else {
+      questionPathToValueMap.put(
+          addressQuestion.getCorrectedPath().toString(), CorrectedAddressState.FAILED.toString());
+    }
+
+    return ImmutableMap.copyOf(questionPathToValueMap);
+  }
+
+  /**
+   * Finds the first {@link ApplicantQuestion} that is an address question type and has address
+   * correction enabled.
+   */
+  public ApplicantQuestion getFirstAddressCorrectionEnabledApplicantQuestion(Block block) {
+    Optional<ApplicantQuestion> applicantQuestionMaybe =
+        block.getQuestions().stream()
+            .filter(
+                applicantQuestion ->
+                    applicantQuestion.getQuestionDefinition().isAddress()
+                        && applicantQuestion.isAddressCorrectionEnabled())
+            .findFirst();
+
+    if (applicantQuestionMaybe.isEmpty()) {
+      throw new RuntimeException(
+          String.format(
+              "Expected to find an address with address correction enabled in block %s, but did"
+                  + " not.",
+              block.getId()));
+    }
+
+    return applicantQuestionMaybe.get();
+  }
```

Comment:
Suggestion (blocking): Now that https://github.com/civiform/civiform/pull/4204 is in, that has the same functionality in app/services/applicant/Block.java as `getAddressQuestionWithCorrectionEnabled`. I don't think this should throw an error if there is no question with address correction enabled.

Reply:
What I should have said is that `getAddressQuestionWithCorrectionEnabled`, which can replace `getFirstAddressCorrectionEnabledApplicantQuestion` here, should not be updated to throw an error since it can also be used to determine if there is an address question with correction enabled (and is used to do so in app/services/applicant/ServiceAreaUpdateResolver.java L66). 

`getCorrectedAddress()` should be updated to use `getAddressQuestionWithCorrectionEnabled` and throw an error if that method returns an empty optional.

Reply:
Since I'm calling getFirstAddressCorrectionEnabledApplicantQuestion in a couple places I removed the logic and am calling block.getgetAddressQuestionWithCorrectionEnabled